src/test/java/com/microsoft/azure/relay/HybridConnectionListenerTest.java

```diff
@@ -29,45 +33,135 @@
+			clientConnection.closeAsync().join();
+			
+			assertTrue("listener should still be open", listener.isOnline());
+			assertFalse("client connection should be closed", clientConnection.isOpen());
+			return rendezvousConnection;
+		}).thenCompose(rendezvousConnection -> {
+			return CompletableFutureUtil.delayAsync(Duration.ofMillis(waitMS), EXECUTOR).whenComplete(($void, ex) -> {
+				assertFalse("rendezvous connection should be closed implicitly when client connection closes", rendezvousConnection.isOpen());
+			});
+		});
+		
+		// Rendezvous initiates closing
+		CompletableFuture<Void> rendezvousClosing = client.createConnectionAsync().thenCombineAsync(listener.acceptConnectionAsync(), (rendezvousConnection, clientConnection) -> {
+			assertTrue("client connection should be open", clientConnection != null && clientConnection.isOpen());
+			assertTrue("rendezvous connection should be open", rendezvousConnection != null && rendezvousConnection.isOpen());
+			rendezvousConnection.closeAsync().join();
+			
+			assertTrue("listener should still be open", listener.isOnline());
+			assertFalse("rendezvous connection should be closed", rendezvousConnection.isOpen());
+			return clientConnection;
+		}).thenCompose(clientConnection -> {
+			return CompletableFutureUtil.delayAsync(Duration.ofMillis(waitMS), EXECUTOR).whenComplete(($void, ex) -> {
+				assertFalse("client connection should be closed implicitly when rendezvous connection closes", clientConnection.isOpen());
+			});
+		});
+		
+		CompletableFuture.allOf(clientClosing, rendezvousClosing).whenComplete(($void, ex) -> {
+			if (ex != null) {
+				fail(ex.getMessage());
+			}	
+			listener.close();
+		}).join();
 	}
 	
 	@Test
-	public void acceptWebSocketConnectionTest() {
-		CompletableFuture<Boolean> checkSocketConnectionTask = new CompletableFuture<Boolean>();
-		CompletableFuture<HybridConnectionChannel> conn = listener.acceptConnectionAsync();
-		CompletableFuture<HybridConnectionChannel> clientConnectionTask = client.createConnectionAsync();
-		conn.thenAccept((connection) -> {
-			checkSocketConnectionTask.complete(true);
-			clientConnectionTask.thenAccept(clientConnection -> clientConnection.closeAsync());
+	public void acceptHandlerTest() throws InterruptedException, ExecutionException, TimeoutException {
+		HybridConnectionListener listener = new HybridConnectionListener(CONNECTION_URI, tokenProvider);
+		AtomicInteger handlerExecuted = new AtomicInteger(0);
+		listener.openAsync().join();
+		
+		// Handler accepts the connection
+		listener.setAcceptHandler(context -> {
+			handlerExecuted.incrementAndGet();
+			return true;
 		});
-		assertTrue("Listener failed to accept connections from sender in webSocket mode.", checkSocketConnectionTask.join());
+		
+		client.createConnectionAsync().whenComplete((clientConnection, ex) -> {
+			assertTrue("clientConnection should have been accepted", clientConnection != null && clientConnection.isOpen());
```

Comment:
>clientConnection [](start = 60, length = 16)

Who closes clientConnection?

Reply:
should be closed implicitly by the listener side rendezvous in the line below.

---
In reply to: [261785389](https://github.com/Azure/azure-relay-java/pull/28#discussion_r261785389) [](ancestors = 261785389)

Reply:
does the listener side of the connection block until the sender side of the connection also calls close?

---
In reply to: [262209295](https://github.com/Azure/azure-relay-java/pull/28#discussion_r262209295) [](ancestors = 262209295,261785389)

Reply:
No it does not block. Closing this connection implicitly is part of the previous test case, so if it doesn't happen it should be caught in the other test.

---
In reply to: [262210594](https://github.com/Azure/azure-relay-java/pull/28#discussion_r262210594) [](ancestors = 262210594,262209295,261785389)